common/persistence/pinot/pinot_visibility_metric_clients_test.go

```diff
@@ -0,0 +1,1220 @@
+			visibilityMgr := NewPinotVisibilityMetricsClient(pinotVisibilityManager, mockMetricClient, logger)
+			metricsClient := visibilityMgr.(*pinotVisibilityMetricsClient)
+
+			// mock behaviors
+			mockMetricClient.On("Scope", mock.Anything, mock.Anything).Return(mockScope).Once()
+			mockScope.On("StartTimer", mock.Anything, mock.Anything).Return(testStopwatch).Once()
+
+			if test.expectedError != nil {
+				mockProducer.On("Publish", mock.Anything, mock.MatchedBy(func(input *indexer.PinotMessage) bool {
+					return true
+				})).Return(fmt.Errorf("error")).Once()
+				mockScope.On("IncCounter", mock.Anything, mock.Anything, mock.Anything).Return().Times(3)
+				err := metricsClient.DeleteWorkflowExecution(context.Background(), test.request)
+				assert.Equal(t, err, test.expectedError)
+			} else {
+				mockProducer.On("Publish", mock.Anything, mock.MatchedBy(func(input *indexer.PinotMessage) bool {
+					return true
+				})).Return(nil).Once()
+				mockScope.On("IncCounter", mock.Anything, mock.Anything, mock.Anything).Return().Once()
+				err := metricsClient.DeleteWorkflowExecution(context.Background(), test.request)
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestMetricClientDeleteUninitializedWorkflowExecution(t *testing.T) {
+	// test non-empty request fields match
+	errorRequest := &p.VisibilityDeleteWorkflowExecutionRequest{}
+
+	request := &p.VisibilityDeleteWorkflowExecutionRequest{}
+
+	tests := map[string]struct {
+		request       *p.VisibilityDeleteWorkflowExecutionRequest
+		expectedError error
+	}{
+		"Case1: error case": {
+			request:       errorRequest,
+			expectedError: fmt.Errorf("error"),
+		},
+		"Case2: normal case": {
+			request:       request,
+			expectedError: nil,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			// create mock clients
+			ctrl := gomock.NewController(t)
+			mockPinotClient := pnt.NewMockGenericClient(ctrl)
+			mockProducer := &mocks.KafkaProducer{}
+			mockMetricClient := &metricsClientMocks.Client{}
+			mockScope := &metricsClientMocks.Scope{}
+
+			// create metricClient
+			logger := testlogger.New(t)
+			mgr := NewPinotVisibilityStore(mockPinotClient, &service.Config{
+				ValidSearchAttributes:  dynamicconfig.GetMapPropertyFn(definition.GetDefaultIndexedKeys()),
+				ESIndexMaxResultWindow: dynamicconfig.GetIntPropertyFn(3),
+			}, mockProducer, testlogger.New(t))
+			visibilityStore := mgr.(*pinotVisibilityStore)
+			pinotVisibilityManager := p.NewVisibilityManagerImpl(visibilityStore, logger)
+			visibilityMgr := NewPinotVisibilityMetricsClient(pinotVisibilityManager, mockMetricClient, logger)
+			metricsClient := visibilityMgr.(*pinotVisibilityMetricsClient)
+
+			// mock behaviors
+			mockMetricClient.On("Scope", mock.Anything, mock.Anything).Return(mockScope).Once()
+			mockScope.On("StartTimer", mock.Anything, mock.Anything).Return(testStopwatch).Once()
+
+			if test.expectedError != nil {
```

Comment:
this is probably ok, but in general if we see that we're having to branch out at all in the table-test Run method that's a sign that we should probably use a different pattern. 

A nice way to do this (I think), is to create an affordance - or a function which takes the mock and sets the expectations per Table-test row. Eg: 

```
          tests := map[string]struct {
		request       *p.VisibilityDeleteWorkflowExecutionRequest
                 producerMockAffordance function(*MockPublisher)
		expectedError error
	}{
		"Case1: error case": {
			request:       errorRequest,
                         producerMockAffordance: func(m MockPublisher){
                             m.Expect().Publish(...).Return(errors.New("an error"))
                         }
			expectedError: fmt.Errorf("error"),
		}
	}


for name, td := range tests
      t.Run(name, func(t testing.T) {
          ... create mocks
         td.producerAffordance(mockProducer)

         // assertions
      }
}

```

Reply:
Good idea. Will make another PR to do this.